# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/vm/ArrayBufferViewObject.h
# Commit: 719bfaded837
# Full Hash: 719bfaded83792edabbeb852a60fe75febf9e65c
# Author: Steve Fink <sfink@mozilla.com>
# Date: 2023-09-21 03:21:32
# Regressor Bug: 1690111
# File Overlap Count: 1
# Description:
#   Bug 1690111 - Implement the ability to pin lengths (prevent changes to length, eg via detaching) for ArrayBuffers and their views. r=jonco
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D131301
# ==============================================================================

diff -r 9497d4dcd737 -r 719bfaded837 js/src/vm/ArrayBufferViewObject.h
--- a/js/src/vm/ArrayBufferViewObject.h	Wed Sep 20 09:41:58 2023 +0000
+++ b/js/src/vm/ArrayBufferViewObject.h	Wed Sep 20 09:41:59 2023 +0000
@@ -24,7 +24,10 @@
 
 class ArrayBufferViewObject : public NativeObject {
  public:
-  // Underlying (Shared)ArrayBufferObject.
+  // Underlying (Shared)ArrayBufferObject. ObjectValue if there is
+  // a buffer. Otherwise, the buffer is implicit because the data
+  // is held inline, and the buffer slot will store the pinned status
+  // (FalseValue or TrueValue).
   static constexpr size_t BUFFER_SLOT = 0;
   static_assert(BUFFER_SLOT == JS_TYPEDARRAYLAYOUT_BUFFER_SLOT,
                 "self-hosted code with burned-in constants must get the "
@@ -121,7 +124,8 @@
     return &obj->as<SharedArrayBufferObject>();
   }
   ArrayBufferObjectMaybeShared* bufferEither() const {
-    JSObject* obj = bufferValue().toObjectOrNull();
+    JSObject* obj =
+        bufferValue().isBoolean() ? nullptr : bufferValue().toObjectOrNull();
     if (!obj) {
       return nullptr;
     }
@@ -146,6 +150,40 @@
     return buffer->isDetached();
   }
 
+  bool isLengthPinned() const {
+    Value buffer = bufferValue();
+    if (buffer.isBoolean()) {
+      return buffer.toBoolean();
+    }
+    if (isSharedMemory()) {
+      return true;
+    }
+    return bufferUnshared()->isLengthPinned();
+  }
+
+  bool pinLength(bool pin) {
+    if (isSharedMemory()) {
+      // Always pinned, cannot change.
+      return false;
+    }
+
+    if (hasBuffer()) {
+      return bufferUnshared()->pinLength(pin);
+    }
+
+    // No ArrayBuffer (data is inline in the view). bufferValue() is a
+    // BooleanValue saying whether the length is currently pinned.
+    MOZ_ASSERT(bufferValue().isBoolean());
+
+    bool wasPinned = bufferValue().toBoolean();
+    if (wasPinned == pin) {
+      return false;
+    }
+
+    setFixedSlot(BUFFER_SLOT, JS::BooleanValue(pin));
+    return true;
+  }
+
   size_t byteOffset() const {
     return size_t(getFixedSlot(BYTEOFFSET_SLOT).toPrivate());
   }